File: ios/Relay/HmacUtil.swift
import Foundation
import CryptoKit
enum HmacUtil {
  static func hex(_ d: Data) -> String { d.map{ String(format:"%02x",$0) }.joined() }
  static func hmacSha256Hex(secret: Data, v: Int, id: String, ts: Int64, dest: String, body: String) -> String {
    let msg = "\(v)|\(id)|\(ts)|\(dest)|\(body)"
    let mac = HMAC<SHA256>.authenticationCode(for: Data(msg.utf8), using: SymmetricKey(data: secret))
    return mac.map{ String(format:"%02x",$0) }.joined()
  }
}

File: android/app/src/main/java/com/example/relay/HmacUtil.kt
package com.example.relay
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec
object HmacUtil {
  fun hex(bytes: ByteArray) = bytes.joinToString(""){ "%02x".format(it) }
  fun hmacSha256Hex(secret: ByteArray, v: Int, id: String, ts: Long, dest: String, body: String): String {
    val mac = Mac.getInstance("HmacSHA256")
    mac.init(SecretKeySpec(secret, "HmacSHA256"))
    val msg = "$v|$id|$ts|$dest|$body"
    return hex(mac.doFinal(msg.toByteArray()))
  }
}
